db/hints/host_filter.cc

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2020 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <string_view>
+#include <boost/algorithm/string.hpp>
+#include "to_string.hh"
+#include "host_filter.hh"
+
+namespace db {
+namespace hints {
+
+host_filter::host_filter(host_filter::enabled_for_all_tag)
+        : _enabled_kind(host_filter::enabled_kind::enabled_for_all) {
+}
+
+host_filter::host_filter(host_filter::disabled_for_all_tag)
+        : _enabled_kind(host_filter::enabled_kind::disabled_for_all) {
+}
+
+host_filter::host_filter(std::unordered_set<sstring> allowed_dcs)
+        : _enabled_kind(allowed_dcs.empty() ? enabled_kind::disabled_for_all : enabled_kind::enabled_selectively)
+        , _dcs(std::move(allowed_dcs)) {
+}
+
+bool host_filter::can_hint_for(locator::snitch_ptr& snitch, gms::inet_address ep) const {
+    switch (_enabled_kind) {
+    case enabled_kind::enabled_for_all:
+        return true;
+    case enabled_kind::enabled_selectively:
+        return _dcs.count(snitch->get_datacenter(ep));
```

Comment:
Please, cache a reference to the snitch_ptr in the `host_filter`. 
The best would be to cache a reference to the object itself but it would be a lot riskier and open a possibility for nasty bugs.

Reply:
`host_filter`  will be moved/copied across shards. I'm not sure if it will be safe to cache a reference to the snitch_ptr, even if we initialize it with `locator::i_endpoint_snitch::get_local_snitch_ptr()` in (move|copy) (constructor|assignment operator).

Reply:
You can make it safe by implementing corresponding copy/move constructors.

Reply:
It looks like we can't do that, as `config` (which holds `host_filter` as a member) seems to be created earlier than the snitch instance. In boost tests, I got a lot of assertions similar to this:
```
row_cache_alloc_stress_test: /home/piodul/code/scylla-3/seastar/include/seastar/core/sharded.hh:715: Service& seastar::sharded<T>::local() [with Service = locator::snitch_ptr]: Assertion `local_is_initialized()' failed.
```

Reply:
I guess it's easier to leave this `snitch` argument then.
Alternatives are worse.